### PR TITLE
http-default-accounts: add iDRAC9

### PR DESCRIPTION
**Pull request by cnotin**
Works fine :)
```
PORT    STATE SERVICE REASON
443/tcp open  https   syn-ack ttl 128
| http-default-accounts: 
|   [Dell iDRAC9] at /
|_    root:calvin
```
Similar to #2095

### CODE REVIEW

**Comment by cnotin** [comment]
Ping @nnposter as you know this very well :)
It might interest you for your fork too!

File: nselib/data/http-default-accounts-fingerprints.lua [hunk 1]
```diff
@@ -1675,6 +1675,40 @@ table.insert(fingerprints, {
...
+  end,
+  login_combos = {
+    {username = "root", password = "calvin"}
+  },
+  login_check = function (host, port, path, user, pass)
+    local headers = {
+                      ["user"]='"'..user..'"',
+                      ["password"]='"'..pass..'"'
+                    }
+    local resp = http_post_simple(host, port, url.absolute(path, "/sysmgmt/2015/bmc/session"),
```
Comment (nnposter) [inline nselib/data/http-default-accounts-fingerprints.lua:1703]: See the note about `url.absolute()` above

File: nselib/data/http-default-accounts-fingerprints.lua [hunk 2]
```diff
@@ -1675,6 +1675,40 @@ table.insert(fingerprints, {
...
+table.insert(fingerprints, {
+  name = "Dell iDRAC9",
+  cpe = "cpe:/o:dell:idrac9_firmware",
+  category = "console",
+  paths = {
+    {path = "/"}
+  },
+  target_check = function (host, port, path, response)
+    -- analyze response for 1st request to "/"
+    if not (response.status == 302 and (response.header["location"] or ""):find("/restgui/start.html$")) then return false end
```
Comment (nnposter) [inline nselib/data/http-default-accounts-fingerprints.lua:1687]: Technically, the dot in the find pattern should be escaped

File: nselib/data/http-default-accounts-fingerprints.lua [hunk 3]
```diff
@@ -1675,6 +1675,40 @@ table.insert(fingerprints, {
...
+
+    -- check with 2nd request to "/restgui/start.html" to be sure
+    local resp = http_get_simple(host, port, url.absolute(path, "/restgui/start.html"))
```
Comment (nnposter) [inline nselib/data/http-default-accounts-fingerprints.lua:1690]: Calling `url.absolute()` with an absolute path as the second argument defeats the purpose of the call. The result will be merely equal to this argument, not its logical concatenation with `path`, which is what you want here.

File: nselib/data/http-default-accounts-fingerprints.lua [hunk 4]
```diff
@@ -1675,6 +1675,40 @@ table.insert(fingerprints, {
...
+  login_check = function (host, port, path, user, pass)
+    local headers = {
+                      ["user"]='"'..user..'"',
+                      ["password"]='"'..pass..'"'
+                    }
+    local resp = http_post_simple(host, port, url.absolute(path, "/sysmgmt/2015/bmc/session"),
+                                 {header=headers})
+    local body = resp.body or ""
+
+    return (resp.status == 201 and body:find('"authResult":0'))
```
[the remark this answers is not in the document]
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua:1707]: I did not observe codes 7 and 9, can you please describe what they refer to?

File: nselib/data/http-default-accounts-fingerprints.lua [hunk 5]
```diff
@@ -1675,6 +1675,40 @@ table.insert(fingerprints, {
...
+            or (resp.status == 503 and body:find('"authResult":6')) -- RAC0218: The maximum number of user sessions is reached
```
[the remark this answers is not in the document]
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua:1708]: I agree with your understanding of code 6, however I recall that I noticed that it only answered this if the authentication was successful (we can suppose that it tries to open a new session, and fails, only if auth succeeds). I'll check again if you prefer.

File: nselib/data/http-default-accounts-fingerprints.lua (hunk 3 above)
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua:1690]: Indeed thx!

**Comment by cnotin** [comment]
Thanks for your detailed feedback @nnposter!
Two comments remain open for now
I'll test your version too ;)

File: nselib/data/http-default-accounts-fingerprints.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua:1707]: > 7 = login success with default credentials.
Src: https://github.com/bmc-toolbox/bmclib/blob/c377a45ac4a6e77d69120e9274b32ba10885c791/providers/dell/idrac9/setupConnections.go#L88

[the remark this answers is not in the document]
Comment (nnposter) [reply nselib/data/http-default-accounts-fingerprints.lua]: 9 appears to be a forced password reset

File: nselib/data/http-default-accounts-fingerprints.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (nnposter) [reply nselib/data/http-default-accounts-fingerprints.lua:1709]: I have tested an iDRAC9 card in this state and it was giving me the same 503/6 regardless of the password.

[the remark this answers is not in the document]
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua:1709]: I just checked and indeed you're right, sorry and thanks for insisting!

File: nselib/data/http-default-accounts-fingerprints.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (cnotin) [reply nselib/data/http-default-accounts-fingerprints.lua]: Good to know! I'll add a comment

**Comment by cnotin** [comment]
What do you think of this? Hope the parenthesis indentation is fine!

**Comment by nnposter** [comment]
The fingerprint has been committed as r37980. Thank you for contributing to nmap!

**Comment by goarblt** [comment]
hello
https://www.evil.com/

**Pull request closed without merging**